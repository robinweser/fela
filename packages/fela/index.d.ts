declare module "fela" {
  import * as CSS from 'csstype';
  import { TRuleType, TKeyframeType, TFontType, TStaticType, TClearType } from 'fela-utils';

  export type TRuleProps = {};
  export type TRule<T = TRuleProps> = (props: T, renderer: IRenderer) => IStyle
  export type TKeyFrame<T = TRuleProps> = (props: T, renderer: IRenderer) => {
    from?: IStyle,
    to?: IStyle,

    [persent: string]: IStyle | undefined;
  };

  type TRendererCreator = (config?: IConfig) => IRenderer;
  type TPlugin = (style: IStyle) => IStyle; //http://fela.js.org/docs/advanced/Plugins.html
  type TEnhancer = (renderer: IRenderer) => IRenderer; //http://fela.js.org/docs/advanced/Enhancers.html

  type TSubscribeMessageType = TRuleType | TKeyframeType | TFontType | TStaticType | TClearType

  interface ISubscribeMessage {
    type: TSubscribeMessageType;
  }
  interface ISubscribeRuleOrStaticObjectMessage extends ISubscribeMessage { static?: boolean; declaration: string; selector: string; media: string; }
  interface ISubscribeKeyframesMessage extends ISubscribeMessage { name: string; keyframe: string; }
  interface ISubscribeFontFaceMessage extends ISubscribeMessage { fontFamily: string; fontFace: string; }
  interface ISubscribeStaticStringMessage extends ISubscribeMessage { css: string; }
  interface ISubscribeClearMessage extends ISubscribeMessage { }

  interface IRenderer {
    renderRule<T = TRuleProps>(rule: TRule<T>, props: T): string
    renderKeyframe<T = TRuleProps>(keyFrame: TKeyFrame<T>, props: T): string;
    renderFont<T = TRuleProps>(family: string, files: Array<string>, props: T): void;
    renderStatic(style: string, selector?: string): void;
    renderStatic(style: IStyle, selector: string): void;
    renderToString(): string;
    subscribe(event: (msg: ISubscribeRuleOrStaticObjectMessage | ISubscribeKeyframesMessage | ISubscribeFontFaceMessage | ISubscribeStaticStringMessage | ISubscribeClearMessage) => void): { unsubscribe: () => void; }
    clear(): void;
  }

  //http://fela.js.org/docs/advanced/RendererConfiguration.html
  interface IConfig {
    plugins?: Array<TPlugin>;
    keyframePrefixes?: Array<string>;
    enhancers?: Array<TEnhancer>;
    mediaQueryOrder?: Array<string>;
    selectorPrefix?: string;
    specificityPrefix?: string;
    filterClassName?: (className: string) => boolean;
    devMode?: boolean;
  }

  export interface IStyle extends CSS.Properties<string | number> {
    // for selectors and pseudo classes use fela-plugin-typescript
  }

  function createRenderer(config?: IConfig): IRenderer;

  function combineRules<A, B>(a: TRule<A>, b: TRule<B>): TRule<A & B>
  function combineRules<A, B, C>(
    a: TRule<A>,
    b: TRule<B>,
    c: TRule<C>,
  ): TRule<A & B & C>
  function combineRules(...rules: Array<TRule>): TRule

  function enhance(...enhancers: Array<TEnhancer>): (rendererCreator: TRendererCreator) => TRendererCreator;
}

declare module "fela-dom" {
  import { IRenderer } from 'fela';

  function render(renderer: IRenderer): void;
  function rehydrate(renderer: IRenderer): void;
  function renderToMarkup(renderer: IRenderer): string;
  function renderToSheetList(renderer: IRenderer): {
    css: string,
    type: 'RULE' | 'KEYFRAME' | 'FONT' | 'STATIC',
    media?: string,
    support?: boolean,
  }[];
}

declare module "fela-tools" {
  import { TRule, TRuleProps, IStyle, IRenderer } from "fela";

  export type TMultiRuleObject<Props = TRuleProps, Styles = {}> = {[key in keyof Styles]: TRule<Props> | IStyle}
  export type TMultiRuleFunction<Props = TRuleProps, Styles = {}> = (props: Props, renderer: IRenderer) => TMultiRuleObject<Props, Styles>
  export type TMultiRule<Props = TRuleProps, Styles = {}> = TMultiRuleObject<Props, Styles> | TMultiRuleFunction<Props, Styles>

  export type TPartialMultiRuleObject<Props = TRuleProps, Styles = {}> = Partial<TMultiRuleObject<Props, Styles>>
  export type TPartialMultiRuleFunction<Props = TRuleProps, Styles = {}> = (props: Props, renderer: IRenderer) => TPartialMultiRuleObject<Props, Styles>
  export type TPartialMultiRule<Props = TRuleProps, Styles = {}> = TPartialMultiRuleObject<Props, Styles> | TPartialMultiRuleFunction<Props, Styles>

  export type TNormalizedMultiRule<Props = TRuleProps, Styles = {}> = (props: Props, renderer: IRenderer) => {[key in keyof Styles]: TRule<Props>}

  function combineMultiRules<A, SA, B, SB>(
    a: TMultiRule<A, SA>,
    b: TMultiRule<B, SB>
  ): TNormalizedMultiRule<A & B, SA & SB>
  function combineMultiRules<A, SA, B, SB, C, SC>(
    a: TMultiRule<A, SA>,
    b: TMultiRule<B, SB>,
    c: TMultiRule<C, SC>,
  ): TNormalizedMultiRule<A & B & C, SA & SB & SC>
  function combineMultiRules(...rules: Array<TMultiRule>): TNormalizedMultiRule

  function mapValueToMediaQuery(
    queryValueMap: { [key: string]: string },
    mapper: ((value: string) => object) | string
  ): object;

  function renderToElement(
    renderer: IRenderer,
    mountNode: { textContent: string },
  ): (() => void);

  function renderToString(
    renderer: IRenderer,
  ): string;
}

/**
 * ENHANCERS
 */
declare module "fela-beautifier" {
  import { TEnhancer } from "fela";

  export default function(options?: object): TEnhancer;
}

declare module "fela-identifier" {
  import { TRule, TEnhancer } from "fela";

  interface Configs {
    prefix?: string;
    generator?: (name?: string, index?: number) => string;
  }

  type Identifier = (name?: string) => TRule & {
    className: string;
    toString(): string;
  };

  export default function(configs?: Configs): TEnhancer & Identifier;
}

declare module "fela-layout-debugger" {
  import { TEnhancer } from "fela";

  interface DebuggerOptions {
    mode?: "outline" | "backgroundColor";
    thickness?: number;
  }

  export default function(options?: DebuggerOptions): TEnhancer;
}

declare module "fela-logger" {
  import { TEnhancer } from "fela";

  interface LoggerOptions {
    logCSS?: boolean;
    formatCSS?: boolean;
  }

  export default function(options?: LoggerOptions): TEnhancer;
}

declare module "fela-monolithic" {
  import { TEnhancer } from "fela";

  interface MonolithicOptions {
    prettySelectors?: boolean;
  }

  export default function(options?: MonolithicOptions): TEnhancer;
}

declare module "fela-perf" {
  import { TEnhancer } from "fela";

  export default function(): TEnhancer;
}

declare module "fela-statistics" {
  import { TEnhancer } from "fela";

  export default function(): TEnhancer;
}

declare module "fela-utils" {
  export type TRuleType = "RULE"
  export type TKeyframeType = "KEYFRAME"
  export type TFontType = "FONT"
  export type TStaticType = "STATIC"
  export type TClearType = "CLEAR"

  export const RULE_TYPE: TRuleType
  export const KEYFRAME_TYPE: TKeyframeType
  export const FONT_TYPE: TFontType
  export const STATIC_TYPE: TStaticType
  export const CLEAR_TYPE: TClearType
}

declare module "fela-sort-media-query-mobile-first" {
  import { TEnhancer } from "fela";

  export default function(): TEnhancer;
}

/**
 * PLUGINS
 */
declare module "fela-plugin-bidi" {
  import { TPlugin } from "fela";

  export default function(flowDirection: 'ltr' | 'rtl'): TPlugin;
}

declare module "fela-plugin-custom-property" {
  import { TPlugin } from "fela";

  interface CustomProperties {
    [property: string]: (value: any) => any,
  }

  export default function(properties: CustomProperties): TPlugin;
}

declare module "fela-plugin-embedded" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-extend" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-fallback-value" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-friendly-pseudo-class" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-important" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-isolation" {
  import { TPlugin } from "fela";

  interface IsolationOptions {
    exclude?: string[];
  }

  export default function(options?: IsolationOptions): TPlugin;
}

declare module "fela-plugin-logger" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-named-keys" {
  import { TPlugin } from "fela";

  interface MediaQueryMap {
    [key: string]: string;
  }

  export default function(mediaQueryMap: MediaQueryMap): TPlugin;
}

declare module "fela-plugin-native-media-query" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-placeholder-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-fullscreen-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-pseudo-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-theme-value" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-rtl" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-simulate" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-unit" {
  import { TPlugin } from "fela";

  export type Unit = "ch" | "em" | "ex" | "rem" | "vh" | "vw" | "vmin" | "vmax" | "px" | "cm" | "mm" | "in" | "pc" | "pt" | "mozmm";

  export interface UnitPerProperty {
    [key: string]: string;
  }

  export default function(
    unit?: Unit,
    unitPerProperty?: UnitPerProperty,
    isUnitlessProperty?: (property: string) => boolean,
  ): TPlugin;
}

declare module "fela-plugin-validator" {
  import { TPlugin } from "fela";

  interface Configs {
    logInvalid?: boolean;
    deleteInvalid?: boolean;
    useCSSLint?: boolean | object;
  }

  export default function(configs?: Configs): TPlugin;
}

/**
 * PRESETS
 */
declare module "fela-preset-web" {
  import { TPlugin } from "fela";
  import { Unit, UnitPerProperty } from "fela-plugin-unit";

  type UnitConfig1 = [Unit]
  type UnitConfig2 = [Unit, UnitPerProperty]
  type UnitConfig3 = [Unit, UnitPerProperty, (property: string) => boolean]

  type UnitConfig = UnitConfig1 | UnitConfig2 | UnitConfig3

  export function createWebPreset({ unit }: { unit?: UnitConfig }): TPlugin[];

  const presets: TPlugin[];
  export default presets;
}

declare module "fela-preset-dev" {
  import { TPlugin } from "fela";

  const presets: TPlugin[];
  export default presets;
}
