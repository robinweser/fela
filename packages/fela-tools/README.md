# fela-tools

A set of useful tools to work with Fela.

<img alt="npm version" src="https://badge.fury.io/js/fela-tools.svg"> <img alt="npm downloads" src="https://img.shields.io/npm/dm/fela-tools.svg"> <a href="https://bundlephobia.com/result?p=fela-tools@latest"><img alt="Bundlephobia" src="https://img.shields.io/bundlephobia/minzip/fela-tools.svg"></a>

## Installation
```sh
yarn add fela-tools
```
You may alternatively use `npm i --save fela-tools`.


## Usage
The API is exported by the `fela-tools` package by default e.g.

```javascript
import { StyleSheet } from 'fela-tools'
```

## API Reference

* [`mapValueToMediaQuery(queryValueMap, mapper)`](docs/mapValueToMediaQuery.md)
* [`renderToString(renderer)`](docs/renderToString.md)
* [StyleSheet](docs/StyleSheet.md)

## License
Fela is licensed under the [MIT License](http://opensource.org/licenses/MIT).<br>
Documentation is licensed under [Creative Commons License](http://creativecommons.org/licenses/by/4.0/).<br>
Created with â™¥ by [@robinweser](http://weser.io) and all the great contributors.
