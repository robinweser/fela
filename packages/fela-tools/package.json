{
  "name": "fela-tools",
  "version": "11.2.0",
  "description": "Useful tools for working with Fela",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "sideEffects": false,
  "files": [
    "LICENSE",
    "README.md",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/robinweser/fela/",
  "keywords": [
    "fela",
    "fela-tool",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "peerDependencies": {
    "fela": "^11.2.0"
  },
  "dependencies": {
    "css-in-js-utils": "^3.0.0",
    "fast-loops": "^1.0.0",
    "fela": "^11.2.0",
    "fela-utils": "^11.2.0"
  },
  "gitHead": "e1a48f5d72623a0f95084dc07998dd85aa175bd1"
}
