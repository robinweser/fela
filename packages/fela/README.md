<h1><img alt="Fela" src="docs/res/logo.png" width="203"/></h1>

Fela is a small, high-performant and framework-agnostic toolbelt to handle state-driven styling in JavaScript.<br>
It is dynamic by design and renders your styles depending on your application state.

It generates atomic CSS and supports all common CSS features such as media queries, pseudo classes, keyframes and font-faces. Fela ships with a powerful plugin API adding e.g. [vendor prefixing](packages/fela-plugin-prefixer) or [fallback value](packages/fela-plugin-fallback-value) support.

Fela can be used with [React](https://github.com/robinweser/fela/tree/master/packages/react-fela) or with any other view library. It even supports [React Native](http://fela.js.org/docs/guides/UsageWithReactNative.html).

<a href="https://bundlephobia.com/result?p=fela@latest"><img alt="Bundlephobia" src="https://img.shields.io/bundlephobia/minzip/fela.svg"></a> <img alt="npm downloads" src="https://img.shields.io/npm/dm/fela.svg"> <a href="https://spectrum.chat/fela?tab=posts"><img alt="Spectrum" src="https://img.shields.io/badge/support-spectrum-brightgreen.svg"></a>

## Support Us
Support Robin Weser's work on Fela and its ecosystem directly via [**Patreon**](https://www.patreon.com/robinweser).

## Installation
```sh
yarn add fela
```
You may alternatively use `npm i --save fela`.

## Benefits
* Predictable Styling
* Scoped Atomic CSS
* Minimal Bundle Size
* No Specificity Issues
* No Naming Conflicts
* Framework-Agnostic
* Huge Ecosystem
* i18n Support

## The Gist
Fela's core principle is to consider [style as a function of state](http://fela.js.org/docs/introduction/Principles.html).<br>
The whole API and all plugins and bindings are built on that idea.<br>
It is reactive and auto-updates once registered to the DOM.<br>

The following example illustrates the key parts of Fela though it only shows the very basics.

```javascript
import { createRenderer } from 'fela'

// a simple style rule is a pure function of state
// that returns an object of style declarations
const rule = state => ({
  textAlign: 'center',
  padding: '5px 10px',
  // directly use the state to compute style values
  fontSize: state.fontSize + 'pt',
  borderRadius: 5,
  // deeply nest media queries and pseudo classes
  ':hover': {
    fontSize: state.fontSize + 2 + 'pt',
    boxShadow: '0 0 2px rgb(70, 70, 70)'
  }
})


const renderer = createRenderer()

// fela generates atomic CSS classes in order to achieve
// maximal style reuse and minimal CSS output
const className = renderer.renderRule(rule, {
  fontSize: 14
}) // =>  a b c d e f
```

The generated CSS output would look like this:
```CSS
.a { text-align: center }
.b { padding: 5px 10px }
.c { font-size: 14pt }
.d { border-radius: 5px }
.e:hover { font-size: 16pt }
.f:hover { box-shadow: 0 0 2px rgb(70, 70, 70) }
```

### Primitive Components
If you're using Fela, you're most likely also using React.<br>
Using the [React bindings](packages/react-fela), you get powerful APIs to create primitive components.<br>

> **Read**: [Usage with React](http://fela.js.org/docs/guides/UsageWithReact.html) for a full guide.

```javascript
import React from 'react'
import { RendererProvider, useFela } from 'react-fela'
import { createRenderer } from 'fela';
import { render } from 'react-dom'

const rule = state => ({
  textAlign: 'center',
  padding: '5px 10px',
  // directly use the state to compute style values
  fontSize: state.fontSize + 'pt',
  borderRadius: 5,
  // deeply nest media queries and pseudo classes
  ':hover': {
    fontSize: state.fontSize + 2 + 'pt',
    boxShadow: '0 0 2px rgb(70, 70, 70)'
  }
})

const Button = ({ children, ...props }) => {
  const { css } = useFela(props)
  
  return <button className={css(rule)}>{children}</button>;
}

const renderer = createRenderer()

render(
  <RendererProvider renderer={renderer}>
    <>
      <Button>Basic Button</Button>
      <Button fontSize={18}>Big Button</Button>
    </>
  </RendererProvider>,
  document.body
)
```

> Check this example on [CodeSandbox](https://codesandbox.io/s/fela-demo-7tsj5)

## Examples
* [Fela + React](http://fela.js.org/docs/introduction/Examples.html#react) ([source](examples/example-react))
    * [React Styleguidist](http://fela.js.org/docs/introduction/Examples.html#styleguidist) ([source](examples/example-with-styleguidist))
    * [React Native](http://fela.js.org/docs/introduction/Examples.html#react-native) ([source](examples/example-react-native))
    * [ReasonReact](https://github.com/astrada/bs-react-fela-examples)
    * [Next](https://github.com/zeit/next.js/tree/master/examples/with-fela)
* [Fela + Preact](http://fela.js.org/docs/introduction/Examples.html#preact) ([source](examples/example-preact))
* [Fela + Inferno](http://fela.js.org/docs/introduction/Examples.html#inferno) ([source](examples/example-inferno))
* [Fela + Angular 2](http://fela.js.org/docs/introduction/Examples.html#angular-2) ([source](examples/example-angular2))
    * [TypeScript](http://fela.js.org/docs/introduction/Examples.html#typescript) ([source](examples/example-angular2-typescript))
* [Fela + HyperScript](https://github.com/ahdinosaur/hyper-fela#example)
* [Fela + Cycle](https://github.com/wcastand/cycle-fela-example)
* [Fela + Next.js](https://github.com/vercel/next.js/tree/master/examples/with-fela)

## Documentation
* [Introduction](http://fela.js.org/docs/Introduction.html)
* [Basics](http://fela.js.org/docs/Basics.html)
* [Advanced](http://fela.js.org/docs/Advanced.html)
* [Usage Guides](http://fela.js.org/docs/UsageGuides.html)
* [Recipes](http://fela.js.org/docs/Recipes.html)
* [API Reference](http://fela.js.org/docs/API.html)
* [Migration Guide](http://fela.js.org/docs/MigrationGuide.html)
* [Changelog](http://fela.js.org/docs/Changelog.html)
* [FAQ](http://fela.js.org/docs/FAQ.html)
* [Feedback](http://fela.js.org/docs/Feedback.html)
* [Thanks](http://fela.js.org/docs/Thanks.html)

## Workshop
If you are coming from CSS and want to learn JavaScript Styling with Fela, there is a full-feature [fela-workshop](https://github.com/tajo/fela-workshop) which demonstrates typical Fela use cases. It teaches all important parts, step by step with simple examples. If you already know other CSS in JS solutions and are just switching to Fela, you might not need to do the whole workshop, but it still provides useful information to get started quickly.

## Talks
* [**CSS in JS: The Good & Bad Parts**](https://www.youtube.com/watch?v=95M-2YzyTno) ([Slides](https://speakerdeck.com/robinweser/css-in-js-the-good-and-bad-parts))<br> - *by [Robin Frischmann](https://twitter.com/robinweser)*
* [**CSS in JS: Patterns**](https://www.webexpo.net/prague2018/talk?id=css-in-js-patterns)<br> - *by [Vojtech Miksu](https://twitter.com/vmiksu)*

## Posts
* [**Style as a Function of State**](https://medium.com/@robinweser/styles-as-functions-of-state-1885627a63f7#.6k6i4kdch)<br> - *by [Robin Frischmann](https://twitter.com/robinweser)*
* [**CSS in JS: The Argument Refined**](https://medium.com/@steida/css-in-js-the-argument-refined-471c7eb83955#.3otvkubq4)<br> - *by [Daniel Steigerwald](https://twitter.com/steida)*
* [**What is Fela?**](https://davidsinclair.io/thoughts/what-is-fela)<br> - *by [David Sinclair](https://davidsinclair.io)*
* [**Choosing a CSS in JS library**](https://gist.github.com/troch/c27c6a8cc47b76755d848c6d1204fdaf#file-choosing-a-css-in-js-library-md)<br> - *by [Thomas Roch](https://twitter.com/tcroch)*
* [**Introducing Fela 6.0**](https://medium.com/felajs/the-future-of-fela-d4dad2efad00)<br> - *by [Robin Frischmann](https://twitter.com/robinweser)*
* [**A journey into CSS and then into CSS-in-JS**](https://www.zeolearn.com/magazine/a-journey-into-css-and-then-into-css-in-js)<br> - *by [Prithvi Raju](https://twitter.com/aga5tya)*
* [**CSS In JS — Future of styling components**](https://we-are.bookmyshow.com/css-in-js-future-of-styling-components-ad315eb5448b)<br> - *by [Manjula Dube](https://twitter.com/manjula_dube)*
* [**Styling Components with React Fela**](https://alligator.io/react/styling-with-react-fela/)<br> - *by [Josh Sherman](https://twitter.com/joshtronic)*
* [**The Future of Fela**](https://medium.com/@robinweser/introducing-fela-6-0-289c84b52dd5)<br> - *by [Robin Frischmann](https://twitter.com/robinweser)*

## Ecosystem
There are tons of useful packages maintained within this repository including plugins, enhancers, bindings and tools that can be used together with Fela. Check the [Ecosystem](http://fela.js.org/docs/introduction/Ecosystem.html) documentation for a quick overview.


## Community
Apart from all the packages managed within this repository, there are many community third-party projects that are worth mentioning:

* [aesthetic](https://github.com/milesj/aesthetic) - React style and theme layer with Fela support
* [base-styling-components](https://github.com/pitr12/base-styling-components) - Abstract Box and Text Components
* [bs-react-fela](https://github.com/astrada/bs-react-fela) - BuckleScript / ReasonReact bindings for Fela
* [catstack](https://github.com/root-systems/catstack) - A modular mad science framework for teams working on production web apps
* [css-in-js-playground](https://github.com/DSchau/css-in-js-playground) - A simple playground for CSS in JS solutions
* [cf-ui](https://github.com/cloudflare/cf-ui) - Cloudflare UI Framework
* [counter-component-with-react-mobx-fela](https://github.com/Mercateo/counter-component-with-react-mobx-fela) - Counter Component using Fela
* [cycle-fela](https://github.com/wcastand/cycle-fela) - Cycle bindings for Fela
* [dogstack](https://github.com/root-systems/dogstack) - A popular-choice grab-bag framework for teams working on production web apps
* [fela-components](https://github.com/arturmuller/fela-components) - Styling library for React and Fela
* [fela-react-helpers](https://github.com/vlad-zhukov/fela-react-helpers) - A set of useful helpers for Fela
* [fela-react-prop](https://github.com/codepunkt/fela-react-prop) - Generate class names for fela style rule and apply them as property on a wrapped component
* [fela-styles-connector](https://github.com/dustin-H/fela-styles-connector) - Simplified react-fela `connect` with auto-bound styles
* [frejya](https://github.com/benoneal/freyja): Pass styles as props to components
* [gatsby-plugin-fela](https://github.com/mmintel/gatsby-plugin-fela) - Integrates fela with [Gatsby](http://gatsbyjs.org)
* [hyper-fela](https://github.com/ahdinosaur/hyper-fela) - HyperScript bindings for Fela
* [htz-frontend](https://github.com/Haaretz/htz-frontend) - Source for Haaretz frontend app(s)
* [kilvin](https://github.com/robinweser/kilvin) - Primitive React Layout Components with Fela
* [olymp](https://github.com/olymp/olymp) - Create and build a next gen app using node, react, cssInJS and other cool stuff
* [preact-fela-simple](https://github.com/pshev/preact-fela-simple) - Super simple Preact bindings for Fela
* [reason-react-starter](https://github.com/drejohnson/reason-react-starter) - A ReasonReact starter kit using Fela
* [storybook-addon-props-fela](https://github.com/Kilix/storybook-addon-props-fela): Document the props of your Fela components in storybook.
* [superslider](https://github.com/adamgiacomelli/superslider) - Slider Component using Fela
* [telaviv](https://github.com/dustin-H/telaviv) - React Universal Rendering
* [vashet](https://github.com/derHowie/vashet) - ClojureScript wrapper for Fela
* [veel](https://github.com/queckezz/veel) - Base react styling components using fela with a design system
* [vue-fela](https://github.com/dustin-H/vue-fela) - Vue bindings for Fela
* [black-box](https://github.com/rocketstation/black-box) - combines behavior, presentation, structure in one place & creates all-in-one components using only JS syntax

## Support
Got a question? Come and join us on [Spectrum](https://spectrum.chat/fela)! <br>
We'd love to help out. We also highly appreciate any feedback.<br>
Don't want to miss any update? Follow us on [Twitter](https://twitter.com/felajs).

## Who's using Fela?

> Your company is using Fela, but is not listed yet? [Add your company / organisation](https://github.com/robinweser/fela/edit/master/README.md#L121)

- [abilis](https://www.abilis.de)
- [Bookmyshow](https://in.bookmyshow.com/events)
- [BdP LV RPS](http://www.bdp-rps.de)
- [Cloudflare](https://www.cloudflare.com)
- [dm-drogerie markt](https://www.dm.de)
- [dmTECH](https://www.dmtech.de)
- [Espressive](https://www.espressive.com)
- [Gazzer](https://www.gazzer.io)
- [HelloFresh](https://www.hellofresh.de)
- [Indoqa](https://www.indoqa.com)
- [Kilix](http://kilix.fr)
- [Lusk](https://lusk.io)
- [MediaFire](https://m.mediafire.com)
- [Medium](https://medium.com)
- [Microsoft](https://microsoft.com)
- [N26](https://n26.com)
- [Net-A-Porter](https://www.net-a-porter.com/gb/en/porter)
- [NinjaConcept](https://www.ninjaconcept.com)
- [Optisure](https://www.optisure.de)
- [Robin Weser](http://weser.io)
- [Rocket Station](http://rstation.io)
- [Root Systems](https://www.rootsystems.nz)
- [Space Between](https://www.spacebetween.co.uk/)
- [V2](https://www.v2.com)
- [Volvo Cars](https://www.volvocars.com)
- [Zendesk](https://www.zendesk.com)

## Contributing

This project exists thanks to all the people who contribute.

We highly appreciate any contribution.<br>
For more information follow the [contribution guide](.github/CONTRIBUTING.md).<br>
Also, please read our [code of conduct](.github/CODE_OF_CONDUCT.md).

## License
Fela is licensed under the [MIT License](http://opensource.org/licenses/MIT).<br>
Documentation is licensed under [Creative Common License](http://creativecommons.org/licenses/by/4.0/).<br>
Created with ♥ by [@robinweser](http://weser.io) and all the great contributors.
