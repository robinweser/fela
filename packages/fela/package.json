{
  "name": "fela",
  "version": "11.4.0",
  "description": "State-Driven Styling in JavaScript",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "sideEffects": false,
  "scripts": {
    "build:umd": "webpack --config webpack.config.js"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**",
    "umd/**"
  ],
  "repository": "https://github.com/robinweser/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "functional styling",
    "functional css",
    "atomic css",
    "state-driven",
    "atomic design",
    "state-driven styling",
    "cssinjs"
  ],
  "author": "Robin Weser",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^3.0.0",
    "csstype": "^2.5.5",
    "fast-loops": "^1.0.0",
    "fela-utils": "^11.4.0",
    "isobject": "^3.0.1"
  },
  "gitHead": "4cc4686d9e16877251ac5c73e161076ec7c113a1"
}
