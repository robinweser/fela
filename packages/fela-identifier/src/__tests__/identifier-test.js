import { createRenderer, combineRules } from 'fela'
import { renderToString } from 'fela-tools'
import createIdentifier from '../index'

describe('Fela identifier enhancer', () => {
  it('should produce rule with custom property', () => {
    const identifier = createIdentifier()
    createRenderer({ enhancers: [identifier] })

    const identifierRule1 = identifier()
    const identifierRule2 = identifier()
    expect([identifierRule1(), identifierRule2()]).toMatchSnapshot()
  })

  it('should be thrown error if the identifier was not used as an enhancer', () => {
    const identifier = createIdentifier()

    expect(() => identifier()).toThrowErrorMatchingSnapshot()
  })

  it('when renderer render identifier rule, the class name should be the same as the rule', () => {
    const identifier = createIdentifier()
    const renderer = createRenderer({ enhancers: [identifier] })

    const identifierRule = identifier()
    expect(renderer.renderRule(identifierRule).trim()).toBe(
      identifierRule.className
    )
  })

  it('identifier rule should have a method "toString" which returns the class name', () => {
    const identifier = createIdentifier()
    createRenderer({ enhancers: [identifier] })

    const identifierRule = identifier()
    expect(identifierRule.toString()).toBe(identifierRule.className)
  })

  it('should be able to fully control the process of generating identifiers', () => {
    const identifier = createIdentifier({
      prefix: 'custom-prefix',
      generator: () => 'custom-id',
    })
    createRenderer({ enhancers: [identifier] })

    const identifierRule = identifier('custom-name')
    expect(identifierRule.className).toBe('custom-prefix-custom-name-custom-id')
  })

  it('identifier should produce new identifiers each time', () => {
    const identifier = createIdentifier()
    createRenderer({ enhancers: [identifier] })

    expect(identifier().className).not.toEqual(identifier().className)
  })

  it('identifier rule should be combined with usual style rules', () => {
    const identifier = createIdentifier()
    const renderer = createRenderer({ enhancers: [identifier] })

    const identifierRule = identifier()
    const styleRule = () => ({
      color: 'red',
      fontSize: '12px',
    })
    const combinedRule = combineRules(identifierRule, styleRule)

    expect(combinedRule({}, renderer)).toMatchSnapshot()
  })

  it('rule can have references to several identifiers', () => {
    const identifier = createIdentifier()
    const renderer = createRenderer({ enhancers: [identifier] })

    const identifierRule1 = identifier()
    const identifierRule2 = identifier()
    const styleRule = () => ({
      color: 'red',
      fontSize: '12px',
    })
    const combinedRule = combineRules(
      identifierRule1,
      identifierRule2,
      styleRule
    )

    expect(renderer.renderRule(combinedRule)).toContain(
      identifierRule1.className
    )
    expect(renderer.renderRule(combinedRule)).toContain(
      identifierRule2.className
    )
    expect(renderer.renderRule(combinedRule)).toMatchSnapshot()
  })

  it('identifier rule should not be contained in the output css', () => {
    const identifier = createIdentifier()
    const renderer = createRenderer({ enhancers: [identifier] })

    const identifierRule = identifier()
    const styleRule = () => ({
      color: 'red',
      fontSize: '12px',
    })
    const combinedRule = combineRules(identifierRule, styleRule)
    renderer.renderRule(combinedRule)

    expect(renderToString(renderer)).toMatchSnapshot()
  })
})
