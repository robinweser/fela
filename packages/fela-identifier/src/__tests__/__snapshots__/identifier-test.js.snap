// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fela identifier enhancer identifier enhancer should not remove meta information in rule 1`] = `
".Component_rule1_zuf8d5 {
  padding: 1;
}
.Component_rule2_137u7ef {
  color: red;
}


<div>
  <span className=Component_rule1_zuf8d5 />
  <span className=Component_rule2_137u7ef />
</div>;
"
`;

exports[`Fela identifier enhancer identifier rule should be combined with usual style rules 1`] = `
{
  "--fela-identifier-0": "",
  "color": "red",
  "fontSize": "12px",
}
`;

exports[`Fela identifier enhancer identifier rule should not be contained in the output css 1`] = `".a{color:red}.b{font-size:12px}"`;

exports[`Fela identifier enhancer rule can have references to several identifiers 1`] = `"fela-identifier-0 fela-identifier-1 a b"`;

exports[`Fela identifier enhancer should be thrown error if the identifier was not used as an enhancer 1`] = `"You are trying to create a new identifier before you used the identifier as an enhancer. Please add the identifier as an enhancer in the configuration of the renderer."`;

exports[`Fela identifier enhancer should produce rule with custom property 1`] = `
[
  {
    "--fela-identifier-0": "",
  },
  {
    "--fela-identifier-1": "",
  },
]
`;
